### PR TITLE
dnd-character: implement exercise v1.0.0

### LINKED ISSUE #1593: dnd-character: implement exercise
- [Exercise description](https://github.com/exercism/problem-specifications/blob/master/exercises/dnd-character/description.md)
- [Canonical data](https://github.com/exercism/problem-specifications/blob/master/exercises/dnd-character/canonical-data.json)

### PR DESCRIPTION
**Pull request by GascaK**
Closes #1593
* DND exercise implemented with corresponding tests and example
* Updated README.md

### CODE REVIEW

File: exercises/dnd-character/dnd_character_test.py [hunk 1]
```diff
@@ -0,0 +1,74 @@
+import unittest
+
+from dnd_character import Character
+
+
+# Tests adapted from 'problem-specifications//canonical-data.json' @ v1.0.0
+
+class DnDCharacterTest(unittest.TestCase):
+    def test_modifier_for_score_3_is_n4(self):
+        self.assertEqual(Character().modifier(3), -4)
```
Comment (cmccandless) [inline exercises/dnd-character/dnd_character_test.py]: `modifier` should probably be separate from `Character`, as it does not use any of `Character`'s members (i.e. it is pure).

File: exercises/dnd-character/dnd_character_test.py [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+        Char = Character()
+        self.assertIn(Char.character['strength'], range(3, 18))
+        self.assertIn(Char.character['dexterity'], range(3, 18))
+        self.assertIn(Char.character['constitution'], range(3, 18))
+        self.assertIn(Char.character['intelligence'], range(3, 18))
+        self.assertIn(Char.character['wisdom'], range(3, 18))
+        self.assertIn(Char.character['charisma'], range(3, 18))
+        self.assertEqual(
+            Char.character['hitpoints'],
+            10 + Char.modifier(Char.character['constitution']))
```
Comment (cmccandless) [inline exercises/dnd-character/dnd_character_test.py]: I think these values should be properties. Wrapping them all in a dict member called (confusingly) `Character.character` is unnecessary.

File: exercises/dnd-character/example.py [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+            'constitution': self.ability(),
+            'intelligence': self.ability(),
+            'wisdom': self.ability(),
+            'charisma': self.ability()
+            }
+        self.character['hitpoints'] = 10 + self.modifier(
+                    self.character['constitution'])
+
+    def ability(self):
+        return random.randint(3, 18)
```
Comment (cmccandless) [inline exercises/dnd-character/example.py]: The intent of the exercise is "to roll 4 dice and sum the highest 3" to generate ability values. Technically, this produces the same result as `randint(3, 18)`... I'm stuck on:

- How essential this difference is in properly holding to the spirit of the exercise?
- How to enforce the "roll 4 dice" method

**Comment by GascaK** [comment]
We can have the ability function return 3 values and sum them in an upper
function? Testing for that can verify if 3 values were returned.

File: exercises/dnd-character/example.py (hunk 3 above)
Comment (cmccandless) [reply exercises/dnd-character/example.py]: > We can have the ability function return 3 values and sum them in an upper
function? Testing for that can verify if 3 values were returned.

I asked this question in the PR where this exercise was implemented. I really like the response I got: https://github.com/exercism/problem-specifications/pull/1397#issuecomment-443716867

To apply this statement to this PR: leave the tests as-is, change the example implementation to match the intent.

```suggestion
		dice_rolls = sorted([random.randint(1, 6) for _ in range(4)])
		return sum(dice_rolls[1:])
```

File: exercises/dnd-character/README.md [hunk 4]
```diff
@@ -0,0 +1,82 @@
+# Acronym
```
Comment (cmccandless) [inline exercises/dnd-character/README.md]: Please regenerate the README using [`configlet`](https://github.com/exercism/configlet)

Comment (GascaK) [reply exercises/dnd-character/README.md]: I configured the configlet tool and have made the necessary changes. Does the configlet tool also edit the config.json or is that done manually? I cannot seem to find the options to change that correctly.

File: exercises/dnd-character/dnd_character_test.py (hunk 1 above)
Comment (GascaK) [reply exercises/dnd-character/dnd_character_test.py]: Deleted the modifier method and created a modifier function in example.py

File: exercises/dnd-character/dnd_character_test.py (hunk 2 above)
Comment (GascaK) [reply exercises/dnd-character/dnd_character_test.py]: Deleted the character dict, and made all values properties.

File: exercises/dnd-character/example.py (hunk 3 above)
Comment (GascaK) [reply exercises/dnd-character/example.py]: Changed the ability method to better exemplify the spirit of the exercise.

File: exercises/dnd-character/README.md (hunk 4 above)
Comment (GascaK) [reply exercises/dnd-character/README.md]: config.json updated manually. Let me know if you agree with the topics section.

File: config.json [hunk 5]
```diff
@@ -1393,6 +1393,17 @@
         "strings"
       ]
     },
+    {
+        "slug": "dnd-character",
+        "uuid": "58625685-b5cf-4e8a-b3aa-bff54da0689d",
+        "core": false,
+        "unlocked_by": null,
```
Comment (cmccandless) [inline config.json]: Since this is the second exercise in this track that deals with `randomness`, I would slot it after `robot-name`:

```suggestion
        "unlocked_by": "robot-name",
```

File: config.json [hunk 6]
```diff
@@ -1393,6 +1393,17 @@
         "strings"
       ]
     },
+    {
+        "slug": "dnd-character",
+        "uuid": "58625685-b5cf-4e8a-b3aa-bff54da0689d",
+        "core": false,
+        "unlocked_by": "robot-name",
```
Comment (cmccandless) [inline config.json]: I made a mistake. Since `robot-name` is not a core exercise, it cannot unlock other exercises. Please revert the change I requested.
```suggestion
        "unlocked_by": null,
```

Comment (cmccandless) [reply config.json]: This is why the Travis build failed. Should be ready to merge after this.

**Comment by cmccandless** [comment]
exercism/problem-specifications#1416 updated the canonical data to 1.1.0. Do you mind updating your PR to match? There was just 1 test case added, but it might change your example implementation.

**Comment by GascaK** [comment]
Sure no prob. Should I create a new pull request or can I just update this one.

**Comment by cmccandless** [comment]
Updating this one will be fine.

File: exercises/dnd-character/dnd_character_test.py [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+        self.assertIn(Character().ability(), range(3, 18))
+
+    def test_random_character_is_valid(self):
+        Char = Character()
+        self.assertIn(Char.strength, range(3, 18))
+        self.assertIn(Char.dexterity, range(3, 18))
+        self.assertIn(Char.constitution, range(3, 18))
+        self.assertIn(Char.intelligence, range(3, 18))
+        self.assertIn(Char.wisdom, range(3, 18))
+        self.assertIn(Char.charisma, range(3, 18))
```
Comment (cmccandless) [inline exercises/dnd-character/dnd_character_test.py]: The second argument of `range()` is exclusive.

```suggestion
        self.assertIn(Char.charisma, range(3, 19))
```

Comment (GascaK) [reply exercises/dnd-character/dnd_character_test.py]: Yup forgot sorry.

**Comment by cmccandless** [comment]
Merged; thanks for working on this!
